### PR TITLE
Django 2.0 & 2.1 support

### PR DESCRIPTION
**Pull request by toffi9**
Before merge of this PR please merge https://github.com/divio/djangocms-text-ckeditor/pull/483

Please @czpython write in review what will be your desired travis matrix for dj21 dj20 and dj111. And also please write how this matrix works because its only travis without tox.

### CODE REVIEW

File: cms/models/pagemodel.py [hunk 1]
```diff
@@ -277,9 +278,10 @@ def _clear_node_cache(self):
         if hasattr(self, '_node_cache'):
             del self._node_cache
 
-        if hasattr(self, 'fields_cache'):
-            # Django >= 2.0
-            self.fields_cache = {}
+        # Django >= 2.0
+        if hasattr(self._state, 'fields_cache'):
```
Comment (czpython) [inline cms/models/pagemodel.py]:
```python
if Page.node.is_cached(self):
    Page.node.delete_cached_value(self)
```

File: cms/tests/test_permmod.py [hunk 2]
```diff
@@ -990,10 +991,12 @@ def test_emulate_admin_index(self):
                 request.user = user
                 # Note, the query count is inflated by doing additional lookups
                 # because there's a site param in the request.
-                with self.assertNumQueries(FuzzyInt(3,4)):
+                with self.assertNumQueries(FuzzyInt(4,5)):
```
Comment (czpython) [inline cms/tests/test_permmod.py]: why is query count going up?

File: cms/appresolver.py [hunk 3]
```diff
@@ -2,20 +2,28 @@
...
 from django.utils.translation import get_language, override
+from django.urls import Resolver404, reverse
 
 from cms.apphook_pool import apphook_pool
 from cms.models.pagemodel import Page
 from cms.utils import get_current_site
 from cms.utils.i18n import get_language_list
 from cms.utils.moderator import use_draft
 
+try:
```
Comment (czpython) [inline cms/appresolver.py]: put this in the utils/compat/django file

File: cms/appresolver.py [hunk 4]
```diff
@@ -248,8 +266,11 @@ def _get_app_patterns(site):
         for lang in hooked_applications[page_id].keys():
             (app_ns, inst_ns), current_patterns, app = hooked_applications[page_id][lang]  # nopyflakes
             if not resolver:
+                regex_pattern = r''
+                if django.VERSION >= (2, 0):
```
Comment (czpython) [inline cms/appresolver.py]: regex_pattern is being overriden, use an if / else or just an inline:
`regex_pattern = RegexPattern(regex_pattern) if django.VERSION >= (2, 0) else r''`

File: cms/utils/permissions.py [hunk 5]
```diff
@@ -203,7 +204,13 @@ def get_page_actions_for_user(user, site):
 
     for perm in page_permissions.iterator():
         # set internal fk cache to our page with loaded ancestors and descendants
-        perm._page_cache = pages_by_id[perm.page_id]
+        if django.VERSION >= (2, 0):
+            if not hasattr(perm._state, 'fields_cache'):
```
Comment (czpython) [inline cms/utils/permissions.py]: `PagePermission.page.set_cached_value(pages_by_id[perm.page_id])`

File: cms/utils/permissions.py [hunk 6]
```diff
@@ -354,7 +361,12 @@ def get_view_restrictions(pages):
 
     for perm in page_permissions:
         # set internal fk cache to our page with loaded ancestors and descendants
-        perm._page_cache = pages_by_id[perm.page_id]
+        if django.VERSION >= (2, 0):
+            if not hasattr(perm._state, 'fields_cache'):
+                perm._state.fields_cache = {}
+            perm._state.fields_cache['page'] = pages_by_id[perm.page_id]
```
Comment (czpython) [inline cms/utils/permissions.py]: `PagePermission.page.set_cached_value(pages_by_id[perm.page_id])`

File: test_requirements/requirements_base.txt [hunk 7]
```diff
@@ -19,3 +19,5 @@ pyenchant
 # required to run the server for integration tests
 djangocms-helper
 mock>=2.0.0
+django-reversion>=2.0
```
[the remark this answers is not in the document]
Comment (toffi9) [reply test_requirements/requirements_base.txt]: it's been in only django111 requirements so I thought now it can be in all dj20 dj21 requirements.

[the remark this answers is not in the document]
Comment (czpython) [reply test_requirements/requirements_base.txt]: best to remove it from all.
it is no longer supported

File: cms/models/pagemodel.py [hunk 8]
```diff
@@ -773,6 +775,7 @@ def copy_with_descendants(self, target_node=None, position=None,
             new_root_node.move(target_node, position)
             new_root_node.refresh_from_db(fields=('path', 'depth'))
 
+        self._clear_node_cache()
```
[the remark this answers is not in the document]
Comment (toffi9) [reply cms/models/pagemodel.py]: Problem was that in dj20+ node cache is in ModelState object, when you `copy.copy` object then also _state was referencing to the same object in copied page and in original page. I changed this to setting on Page.copy new instance of ModelState. In new commit you will see this changes.

**Comment by toffi9** [comment]
Still need to resolve conflicts and provide CI config

File: cms/management/commands/subcommands/base.py [hunk 9]
```diff
@@ -58,18 +62,22 @@ def add_arguments(self, parser):
...
-                    description=instance.help_string
+                    name=instance.command_name, help=instance.help_string,
+                    description=instance.help_string, **kwargs
                 )
 
                 add_builtin_arguments(parser=parser_sub)
                 instance.add_arguments(parser_sub)
+                stealth_options.extend({action.dest for action in parser_sub._actions})
                 self.instances[command] = instance
+            self.stealth_options = tuple(set(stealth_options))
```
Comment (czpython) [inline cms/management/commands/subcommands/base.py]: why is this casted to a set and then a tuple?

File: cms/admin/pageadmin.py [hunk 10]
```diff
@@ -442,8 +443,20 @@ def delete_view(self, request, object_id, extra_context=None):
         # Populate deleted_objects, a data structure of all related objects that
         # will also be deleted.
         objs = [obj] + list(obj.get_descendant_pages())
+
+        get_deleted_objects_additional_kwargs = {}
```
Comment (czpython) [inline cms/admin/pageadmin.py]: remove this line

File: cms/migrations/0022_auto_20180620_1551.py [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+        ('cms', '0021_auto_20180507_1432'),
+    ]
+
+    operations = [
+        migrations.AlterModelOptions(
+            name='page',
+            options={'default_permissions': ('add', 'change', 'delete'), 'permissions': (('view_page', 'Can view page'), ('publish_page', 'Can publish page'), ('edit_static_placeholder', 'Can edit static placeholders')), 'verbose_name': 'page', 'verbose_name_plural': 'pages'},
+        ),
+    ]
+    if not DJANGO_2_0:
```
Comment (czpython) [inline cms/migrations/0022_auto_20180620_1551.py]: why is this needed on django >= 2.1?

File: cms/admin/permissionadmin.py [hunk 12]
```diff
@@ -38,6 +38,27 @@ class PagePermissionInlineAdmin(TabularInline):
     extra = 0  # edit page load time boost
     show_with_view_permissions = False
 
+    def has_change_permission(self, request, obj=None):
```
Comment (czpython) [inline cms/admin/permissionadmin.py]: there's no pagepermission permission.
did you get a failing test?

File: cms/management/commands/subcommands/base.py (hunk 9 above)
Comment (toffi9) [reply cms/management/commands/subcommands/base.py]: `set` because I wanted only unique entries and `tuple` to back to original type

File: cms/models/pagemodel.py [hunk 13]
```diff
@@ -663,6 +666,7 @@ def copy(self, site, parent_node=None, language=None,
             parent_page = None
 
         new_page = copy.copy(self)
+        new_page._state = ModelState()
```
[the remark this answers is not in the document]
Comment (toffi9) [reply cms/models/pagemodel.py]: Problem is that from dj20 fields caching is in _state object when we copy page like above, then also a _state object is copied, and it's referencing to the same object as original page, and it generated failures in tests. I resolved this issue by creating new instance of this object, this is done the same way as in django (https://github.com/django/django/blob/1.11/django/db/models/base.py#L481).

File: cms/migrations/0022_auto_20180620_1551.py (hunk 11 above)
Comment (toffi9) [reply cms/migrations/0022_auto_20180620_1551.py]: When running tests against dj21 it's showing missing migration

File: cms/admin/permissionadmin.py (hunk 12 above)
Comment (toffi9) [reply cms/admin/permissionadmin.py]: No, all test passes. This is needed only for django 21 but it works also on dj111 and d20. It's needed because of the changes that django 21 made regarding admin. https://docs.djangoproject.com/en/2.1/releases/2.1/#custom-admin-forms-need-to-take-the-view-only-case-into-account. And in code especially: https://github.com/olivierdalang/django/commit/63b62af169f6fc02182e35a058dcac78bdaff049#diff-3c42de3e53aba87b32c494f995a728dfR2024.

Comment (toffi9) [reply cms/admin/permissionadmin.py]: https://github.com/django/django/blob/2.1b1/django/contrib/admin/options.py#L2060 this is causing the need for this code. In django 2.1 they do this and if user dont have sufficient permissions fields are removed. And it's causing test errors. It's somehow described here: https://docs.djangoproject.com/en/2.1/releases/2.1/#custom-admin-forms-need-to-take-the-view-only-case-into-account. Solution showed by django docs it's for very simple case, and its not working for us.

File: cms/admin/permissionadmin.py [hunk 14]
```diff
@@ -38,6 +38,27 @@ class PagePermissionInlineAdmin(TabularInline):
...
+        site = self.get_site(request)
+        if not obj:
+            return permissions.has_global_permission(request.user, site, 'can_change_pagepermission')
+        return page_permissions.user_can_change_page_permissions(request.user, page=obj, site=site)
```
Comment (toffi9) [inline cms/admin/permissionadmin.py]: `obj` here is always `Page` model, not `PagePermission` seems like django works like that in `ModelAdmin` in `changeform_view>>_changeform_view>>_create_formsets>>get_formsets_with_inlines>>get_inline_instances(self, request, obj=None)`

File: cms/admin/permissionadmin.py (hunk 12 above)
Comment (toffi9) [reply cms/admin/permissionadmin.py]: @czpython `obj` here is always `Page` model, not `PagePermission` seems like django works like that in `ModelAdmin` in `changeform_view>>_changeform_view>>_create_formsets>>get_formsets_with_inlines>>get_inline_instances(self, request, obj=None)`
